File: pc-sync-tool/src/v2_manifest.py
from __future__ import annotations

import json
import re
from datetime import datetime, timezone
from typing import Any, Mapping

from v2_contract import SCHEMA_VERSION, SOURCE_TYPE, SOFT_LIMITS, snapshot_download_path, validate_snapshot_id


FIELDS = frozenset(
    {
        "ok",
        "schemaVersion",
        "snapshotId",
        "sourceType",
        "createdAtUtc",
        "sizeBytes",
        "sha256",
        "minimumAppVersion",
        "productCount",
        "categoryCount",
        "unitCount",
        "promotionCandidateCount",
        "verifiedPromotionCount",
        "validationIssueCount",
        "downloadPath",
    }
)
_SHA256 = re.compile(r"^[0-9a-f]{64}$")
_INT32_MAX = 2_147_483_647


def validate_v2_manifest(data: Mapping[str, Any], encoded_size: int | None = None) -> dict[str, Any]:
    if not isinstance(data, Mapping) or set(data) != FIELDS:
        raise ValueError("v2 manifest fields do not match the frozen schema")
    if data["ok"] is not True:
        raise ValueError("ok must be JSON boolean true")
    if data["schemaVersion"] != SCHEMA_VERSION:
        raise ValueError("unsupported schemaVersion")
    snapshot = validate_snapshot_id(data["snapshotId"])
    if data["sourceType"] != SOURCE_TYPE:
        raise ValueError("invalid sourceType")
    _validate_utc_instant(data["createdAtUtc"])
    if not isinstance(data["sha256"], str) or not _SHA256.fullmatch(data["sha256"]):
        raise ValueError("sha256 must be lowercase hexadecimal")

    integer_fields = (
        "sizeBytes",
        "minimumAppVersion",
        "productCount",
        "categoryCount",
        "unitCount",
        "promotionCandidateCount",
        "verifiedPromotionCount",
        "validationIssueCount",
    )
    for field in integer_fields:
        value = data[field]
        if isinstance(value, bool) or not isinstance(value, int) or value < 0:
            raise ValueError(f"{field} must be a non-negative JSON integer")
    if data["minimumAppVersion"] <= 0:
        raise ValueError("minimumAppVersion must be a positive versionCode")
    for field in ("categoryCount", "unitCount"):
        if data[field] > _INT32_MAX:
            raise ValueError(f"{field} exceeds the frozen int32 manifest limit")
    if data["minimumAppVersion"] > _INT32_MAX:
        raise ValueError("minimumAppVersion exceeds the frozen int32 manifest limit")
    if data["sizeBytes"] > SOFT_LIMITS["snapshotBytes"]:
        raise ValueError("snapshot exceeds soft limit")
    if data["productCount"] > SOFT_LIMITS["productCount"]:
        raise ValueError("product count exceeds soft limit")
    if data["promotionCandidateCount"] > SOFT_LIMITS["promotionCandidateCount"]:
        raise ValueError("promotion candidate count exceeds soft limit")
    if data["validationIssueCount"] > SOFT_LIMITS["validationIssueCount"]:
        raise ValueError("validation issue count exceeds soft limit")
    if data["verifiedPromotionCount"] > data["promotionCandidateCount"]:
        raise ValueError("verified promotion count exceeds candidate count")
    if data["downloadPath"] != snapshot_download_path(snapshot):
        raise ValueError("downloadPath does not match snapshotId")
    actual_size = encoded_size
    if actual_size is None:
        actual_size = len(json.dumps(dict(data), separators=(",", ":"), ensure_ascii=False).encode("utf-8"))
    if actual_size > SOFT_LIMITS["manifestBytes"]:
        raise ValueError("manifest exceeds soft limit")
    return dict(data)


def _validate_utc_instant(value: Any) -> None:
    if not isinstance(value, str) or not value.endswith("Z"):
        raise ValueError("createdAtUtc must be a UTC instant")
    try:
        parsed = datetime.fromisoformat(value[:-1] + "+00:00")
    except ValueError as exc:
        raise ValueError("createdAtUtc must be a UTC instant") from exc
    if parsed.utcoffset() != timezone.utc.utcoffset(None):
        raise ValueError("createdAtUtc must be UTC")

File: core/src/main/java/com/espsa/mobilepos/core/model/V2Manifest.java
package com.espsa.mobilepos.core.model;

import java.time.Instant;
import java.util.regex.Pattern;

public final class V2Manifest {
    private static final Pattern SHA256 = Pattern.compile("^[0-9a-f]{64}$");

    private final String snapshotId;
    private final String createdAtUtc;
    private final long sizeBytes;
    private final String sha256;
    private final int minimumAppVersion;
    private final int productCount;
    private final int categoryCount;
    private final int unitCount;
    private final int promotionCandidateCount;
    private final int verifiedPromotionCount;
    private final int validationIssueCount;
    private final String downloadPath;

    public V2Manifest(
            boolean ok, int schemaVersion, String snapshotId, String sourceType,
            String createdAtUtc, long sizeBytes, String sha256, long minimumAppVersion,
            int productCount, long categoryCount, long unitCount, int promotionCandidateCount,
            int verifiedPromotionCount, int validationIssueCount, String downloadPath
    ) {
        if (!ok || schemaVersion != V2Contract.SCHEMA_VERSION || !V2Contract.SOURCE_TYPE.equals(sourceType)) {
            throw new IllegalArgumentException("Invalid manifest identity fields");
        }
        this.snapshotId = V2Contract.validateSnapshotId(snapshotId);
        validateUtc(createdAtUtc);
        if (sizeBytes < 0 || sizeBytes > V2Contract.SNAPSHOT_SOFT_BYTES
                || minimumAppVersion < 1 || minimumAppVersion > Integer.MAX_VALUE
                || productCount < 0 || productCount > V2Contract.PRODUCT_SOFT_COUNT
                || categoryCount < 0 || categoryCount > Integer.MAX_VALUE
                || unitCount < 0 || unitCount > Integer.MAX_VALUE
                || promotionCandidateCount < 0 || promotionCandidateCount > V2Contract.PROMOTION_SOFT_COUNT
                || verifiedPromotionCount < 0 || verifiedPromotionCount > promotionCandidateCount
                || validationIssueCount < 0 || validationIssueCount > V2Contract.ISSUE_SOFT_COUNT) {
            throw new IllegalArgumentException("Manifest numeric field is outside the contract");
        }
        if (sha256 == null || !SHA256.matcher(sha256).matches()) {
            throw new IllegalArgumentException("Invalid SHA-256");
        }
        if (!V2Contract.downloadPath(snapshotId).equals(downloadPath)) {
            throw new IllegalArgumentException("downloadPath does not match snapshotId");
        }
        this.createdAtUtc = createdAtUtc;
        this.sizeBytes = sizeBytes;
        this.sha256 = sha256;
        this.minimumAppVersion = (int) minimumAppVersion;
        this.productCount = productCount;
        this.categoryCount = (int) categoryCount;
        this.unitCount = (int) unitCount;
        this.promotionCandidateCount = promotionCandidateCount;
        this.verifiedPromotionCount = verifiedPromotionCount;
        this.validationIssueCount = validationIssueCount;
        this.downloadPath = downloadPath;
    }

    public String snapshotId() { return snapshotId; }
    public String createdAtUtc() { return createdAtUtc; }
    public long sizeBytes() { return sizeBytes; }
    public String sha256() { return sha256; }
    public int minimumAppVersion() { return minimumAppVersion; }
    public int productCount() { return productCount; }
    public int categoryCount() { return categoryCount; }
    public int unitCount() { return unitCount; }
    public int promotionCandidateCount() { return promotionCandidateCount; }
    public int verifiedPromotionCount() { return verifiedPromotionCount; }
    public int validationIssueCount() { return validationIssueCount; }
    public String downloadPath() { return downloadPath; }

    private static void validateUtc(String value) {
        if (value == null || !value.endsWith("Z")) {
            throw new IllegalArgumentException("createdAtUtc must be UTC");
        }
        try {
            Instant.parse(value);
        } catch (RuntimeException exception) {
            throw new IllegalArgumentException("createdAtUtc must be a UTC instant", exception);
        }
    }
}
